### PR TITLE
Reset password, state layer.

### PR DESCRIPTION
**Pull request by anastasiamac**
## Description of change

This is part of the functionality to provide controller admins with ability to reset user passwords/re-issue login tokens.

## QA steps

This is only one the layers, there is no full-stack funcationaliy yet. All unit tests pass.

## Documentation changes

n/a
This is an internal addition atm.

## Bug reference

Initial part for https://bugs.launchpad.net/juju/+bug/1657187

### CODE REVIEW

File: state/user_test.go [hunk 1]
```diff
@@ -408,3 +408,53 @@ func (s *UserSuite) TestSetPasswordClearsSecretKey(c *gc.C) {
...
+	c.Assert(err, jc.ErrorIsNil)
+
+	err = u.SetPassword("anything")
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(u.PasswordValid("anything"), jc.IsTrue)
+	c.Assert(u.SecretKey(), gc.IsNil)
+
+	key, err := u.ResetPassword()
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(u.SecretKey(), gc.DeepEquals, key)
```
Comment (axw) [inline state/user_test.go]: also check that PasswordValid("anything") fails here

File: state/user.go [hunk 2]
```diff
@@ -543,6 +540,61 @@ func (u *User) ensureNotDeleted() error {
...
+// It generates and returns a new user secret key.
+// This must be an active user.
+func (u *User) ResetPassword() ([]byte, error) {
+	var key []byte
+	buildTxn := func(attempt int) ([]txn.Op, error) {
+		if err := u.ensureNotDeleted(); err != nil {
+			return nil, errors.Trace(err)
+		}
+		if u.IsDisabled() {
+			return nil, fmt.Errorf("user deactivated")
```
Comment (anastasiamac) [inline state/user.go]: I am not convinced that this needs further clarification. The overall error that will come from here will be `cannot reset password for user "bob": user deactivated`.

File: state/user.go [hunk 3]
```diff
@@ -543,6 +540,61 @@ func (u *User) ensureNotDeleted() error {
 	return nil
 }
 
+// ResetPassword cleans up password related field.
```
Comment (anastasiamac) [inline state/user.go]: updated

File: state/user.go (hunk 2 above)
Comment (axw) [inline state/user.go]: Agreed, I missed the fact that there's additional context below.

File: state/user.go [hunk 4]
```diff
@@ -543,6 +540,61 @@ func (u *User) ensureNotDeleted() error {
...
+			C:      usersC,
+			Id:     u.Name(),
+			Assert: txn.DocExists,
+			Update: update,
+		}}, nil
+	}
+	if err := u.st.db().Run(buildTxn); err != nil {
+		return nil, errors.Annotatef(err, "cannot reset password for user %q", u.Name())
+	}
+	u.doc.SecretKey = key
```
Comment (anastasiamac) [inline state/user.go]: good point! done :D

File: state/user.go [hunk 5]
```diff
@@ -543,6 +540,61 @@ func (u *User) ensureNotDeleted() error {
...
+	return key, nil
+}
+
+// generateSecretKey generates a random, 32-byte secret key. This can be used
```
Comment (anastasiamac) [inline state/user.go]: done

File: state/user_test.go (hunk 1 above)
Comment (anastasiamac) [inline state/user_test.go]: added check.

**Comment by anastasiamac** [comment]
$$merge$$

**Pull request merged**